### ISSUE TITLE
Unexpected compilation error for dynamic logical operator

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#27136]
```
struct C
{
    void M(C? a, dynamic b, dynamic result)
/*<bind>*/{
        result = a && b;
    }/*</bind>*/

    public static bool operator true(C? c) => throw null;
    public static bool operator false(C? c) => throw null;
}
```

Observed:
```
                // file.cs(6,18): error CS7083: Expression must be implicitly convertible to Boolean or its type 'C?' must define operator 'false'.
                //         result = a && b;
                Diagnostic(ErrorCode.ERR_InvalidDynamicCondition, "a").WithArguments("C?", "false").WithLocation(6, 18)
```

Expected:
Should succeed, there is a required operator.
